common/lib/xmodule/xmodule/capa_base.py

```diff
@@ -85,224 +105,44 @@ def default(self, obj):  # pylint: disable=method-hidden
+            return due_date + self.graceperiod
         else:
-            self.close_date = due_date
+            return due_date
 
+    def get_seed(self):
+        """
+        Generate the seed if not set and return it.
+        """
         if self.seed is None:
             self.choose_new_seed()
+        return self.seed
 
-        # Need the problem location in openendedresponse to send out.  Adding
-        # it to the system here seems like the least clunky way to get it
-        # there.
-        self.runtime.set('location', text_type(self.location))
-
+    @cached_property
+    def lcp(self):
         try:
-            # TODO (vshnayder): move as much as possible of this work and error
-            # checking to descriptor load time
-            self.lcp = self.new_lcp(self.get_state_for_lcp())
-
-            # At this point, we need to persist the randomization seed
-            # so that when the problem is re-loaded (to check/view/save)
-            # it stays the same.
-            # However, we do not want to write to the database
-            # every time the module is loaded.
-            # So we set the seed ONLY when there is not one set already
-            if self.seed is None:
-                self.seed = self.lcp.seed
-
+            lcp = self.new_lcp(self.get_state_for_lcp())
         except Exception as err:  # pylint: disable=broad-except
             msg = u'cannot create LoncapaProblem {loc}: {err}'.format(
                 loc=text_type(self.location), err=err)
-            # TODO (vshnayder): do modules need error handlers too?
-            # We shouldn't be switching on DEBUG.
-            if self.runtime.DEBUG:
-                log.warning(msg)
-                # TODO (vshnayder): This logic should be general, not here--and may
-                # want to preserve the data instead of replacing it.
-                # e.g. in the CMS
-                msg = HTML(u'<p>{msg}</p>').format(msg=msg)
-                msg += HTML(u'<p><pre>{tb}</pre></p>').format(
-                    # just the traceback, no message - it is already present above
-                    tb=u''.join(
-                        ['Traceback (most recent call last):\n'] +
-                        traceback.format_tb(sys.exc_info()[2])
-                    )
-                )
-                # create a dummy problem with error message instead of failing
-                problem_text = (
-                    HTML(u'<problem><text><span class="inline-error">'
-                         u'Problem {url} has an error:</span>{msg}</text></problem>').format(
-                        url=text_type(self.location),
-                        msg=msg,
-                    )
-                )
-                self.lcp = self.new_lcp(self.get_state_for_lcp(), text=problem_text)
-            else:
-                # add extra info and raise
-                raise Exception(msg), None, sys.exc_info()[2]
-
-            self.set_state_from_lcp()
```

Comment:
This line only got executed if `self.runtime.DEBUG` was `True` so no longer needed.